refactor(WebWorker): Use the new generic bootstrap.

jteplitz opened the pull request:
BREAKING CHANGE:

You can no longer bootstrap a WebWorker or Isolate using `bootstrap` or `bootstrapWebWorker`. Instead you have to do the following:

In TypeScript:

``` TypeScript
// index.js
import {WORKER_RENDER_PLATFORM, WORKER_RENDER_APPLICATION, WORKER_SCRIPT} from "angular2/platforms/worker_render";
import {platform} from "angular2/platform";

platform([WORKER_RENDER_PLATFORM])
.application([WORKER_RENDER_APPLICATION, new Provider(WORKER_SCRIPT, {useValue: "loader.js"});
```

``` JavaScript
// loader.js
importScripts("https://cdnjs.cloudflare.com/ajax/libs/es6-shim/0.33.3/es6-shim.js", "https://jspm.io/system@0.16.js", "angular2/web_worker/worker.js");
System.import("app");
```

``` TypeScript
// app.ts
import {Component, View} from "angular2/core";
import {WORKER_APP_PLATFORM, setupWebWorker} from "angular2/platforms/worker_app";
import {platform} from "angular2/platform";

@Component({
  selector: "hello-world"
})
@View({
  template: "<h1>Hello {{name}}</h1>
})
export class HelloWorld {
  name: string = "Jane";
}

platform([WORKER_APP_PLATFORM])
.asyncApplication(setupWebWorker, optionalProviders?)
.then((ref) => ref.bootstrap(RootComponent));
```

In Dart:

``` Dart
// index.dart
import "angular2/platform.dart";
import "angular2/platforms/worker_render.dart";

main() {
  platform([WORKER_RENDER_PLATFORM])
  .asyncApplication(initIsolate("my_worker.dart"));
}
```

``` Dart
// background_index.dart
import "angular2/platform.dart";
import "angular2/platforms/worker_app.dart";
import "package:angular2/src/core/reflection/reflection.dart";
import "package:angular2/src/core/reflection/reflection_capabilities.dart";

@Component(
  selector: "hello-world"
)
@View(
  template: "<h1>Hello {{name}}</h1>"
)
class HelloWorld {
  String name = "Jane";
}

main(List<String> args, SendPort replyTo) {
  reflector.reflectionCapabilities = new ReflectionCapabilities();
  platform([WORKER_APP_PLATFORM])
    .asyncApplication(setupIsolate(replyTo))
      .then((ref) => ref.bootstrap(RootComponent));
}

```

You should no longer import from the `angular2/web_worker/worker` and `angular2/web_worker/ui` paths. Instead you can now import directly from core, directives, etc..

The WebWorkerApplication class has been removed. If you want to use ServiceMessageBroker or ClientMessageBroker on the render thread, you must inject their factories via DI.
If you need to use the MessageBus on the render you must also obtain it through DI.

closes #3277

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -39,8 +41,11 @@ To bootstrap Hello World in a WebWorker we do the following in TypeScript
-import {bootstrap} from "angular2/web_worker/ui";
-bootstrap("loader.js");
+import {WORKER_RENDER_PLATFORM, WORKER_RENDER_APPLICATION, WORKER_SCRIPT} from "angular2/platforms/worker_render";
+import {platform} from "angular2/platform";
```
platform is exported by `angular2/core`

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -49,7 +54,10 @@ System.import("app");
-import {Component, View, bootstrapWebWorker} from "angular2/web_worker/worker";
+import {Component, View} from "angular2/core";
+import {WORKER_APP_PLATFORM, setupWebWorker} from "angular2/platforms/worker_app";
+import {platform} from "angular2/platform";
```
platform is exported by `angular2/core`

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -60,21 +68,19 @@ export class HelloWorld {
-these URLs instead of `angular2/angular2` we can statically ensure that our app does not reference a type that
-doesn't exist in the context it's mean to execute in. For example, if we tried to import DomRenderer in the Worker
-or NgFor on the UI we would get a compiler error.
 * The UI loads angular from the file `angular2/web_worker/ui.js` and the Worker loads angular from
```
Is the following still accurate? `The UI loads angular from the file `angular2/web_worker/ui.js`

jteplitz602 replied:
Yes, that's still the name of the bundle file.

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -202,12 +213,18 @@ can make it easy to catch tricky browser compatibility issues.
-var instance = bootstrap("loader.js");
-var bus = instance.bus;
+import {WORKER_RENDER_PLATFORM, WORKER_RENDER_APPLICATION, WORKER_SCRIPT} from "angular2/platforms/worker_render";
+import {platform} from "angular2/platform";
```
angular2/core

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -226,11 +243,15 @@ Once the channel has been initialized either side can use the `from` and `to` me
-var instance = bootstrap("loader.js");
-var bus = instance.bus;
+import {WORKER_RENDER_PLATFORM, WORKER_RENDER_APPLICATION, WORKER_SCRIPT} from "angular2/platforms/worker_render";
+import {platform} from "angular2/platform";
```
angular2/core

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -253,10 +274,16 @@ This example is nearly identical in Dart, and is included below for reference:
-  var bus = instance.bus;
-  bus.initChannel("My Custom Channel");
-  bus.to("My Custom Channel").add("hello from the UI");
+  import "angular2/platform.dart";
```
core

vsavkin commented on modules/angular2/docs/web_workers/web_workers.md:
```diff
@@ -338,54 +365,98 @@ class JsonMessageBusSource extends GenericMessageBuSource {
+    WORKER_RENDER_APP_COMMON,
+    initializeGenericWorkerRenderer
+} from 'angular2/platform/worker_render';
+import {MessageBus} from 'angular2/web_worker/ui';
```
I think `platform/worker_render` should export MessageBus instead of  `web_worker/ui`. I think that we should remove web_worker ui altogether. 

The purpose of platform isn't just to provide two lists of providers, but to export all platform-specific symbols too.

vsavkin replied:
I think that `web_worker/worker` should be removed too.

jteplitz602 replied:
I agree. However, `web_worker/worker.ts` and `web_worker/ui.ts` are the files that we're using to specify what belongs in the webworker bundles. Is there a way to specify the bundle contents without creating a new import path?

jteplitz602 replied:
Or should we make the new bundles `platforms/worker_app.js` and `platforms/worker_render.js` and kill off the `web_worker` folder?

vsavkin commented on modules/angular2/src/core/application_ref.ts:
```diff
@@ -217,12 +217,15 @@ export class PlatformRef_ extends PlatformRef {
     var completer = PromiseWrapper.completer();
     zone.run(() => {
       PromiseWrapper.then(bindingFn(zone), (providers: Array<Type | Provider | any[]>) => {
+        if (isPresent(additionalProviders)) {
```
Since this is in core, can you add a web-worker-independent unit test for this change.

jteplitz602 replied:
done

jteplitz committed:
refactor(WebWorker): Use the new generic bootstrap.

BREAKING CHANGE:

You can no longer bootstrap a WebWorker or Isolate using `bootstrap` or `bootstrapWebWorker`. Instead you have to do the following:

In TypeScript:
```TypeScript
// index.js
import {WORKER_RENDER_PLATFORM, WORKER_RENDER_APPLICATION, WORKER_SCRIPT} from "angular2/platforms/worker_render";
import {platform} from "angular2/platform";

platform([WORKER_RENDER_PLATFORM])
.application([WORKER_RENDER_APPLICATION, new Provider(WORKER_SCRIPT, {useValue: "loader.js"});
```
```JavaScript
// loader.js
importScripts("https://cdnjs.cloudflare.com/ajax/libs/es6-shim/0.33.3/es6-shim.js", "https://jspm.io/system@0.16.js", "angular2/web_worker/worker.js");
System.import("app");
```
```TypeScript
// app.ts
import {Component, View} from "angular2/core";
import {WORKER_APP_PLATFORM, setupWebWorker} from "angular2/platforms/worker_app";
import {platform} from "angular2/platform";

@Component({
  selector: "hello-world"
})
@View({
  template: "<h1>Hello {{name}}</h1>
})
export class HelloWorld {
  name: string = "Jane";
}

platform([WORKER_APP_PLATFORM])
.asyncApplication(setupWebWorker, optionalProviders?)
.then((ref) => ref.bootstrap(RootComponent));
```

In Dart:
```Dart
// index.dart
import "angular2/platform.dart";
import "angular2/platforms/worker_render.dart";

main() {
  platform([WORKER_RENDER_PLATFORM])
  .asyncApplication(initIsolate("my_worker.dart"));
}
```
```Dart
// background_index.dart
import "angular2/platform.dart";
import "angular2/platforms/worker_app.dart";
import "package:angular2/src/core/reflection/reflection.dart";
import "package:angular2/src/core/reflection/reflection_capabilities.dart";

@Component(
  selector: "hello-world"
)
@View(
  template: "<h1>Hello {{name}}</h1>"
)
class HelloWorld {
  String name = "Jane";
}

main(List<String> args, SendPort replyTo) {
  reflector.reflectionCapabilities = new ReflectionCapabilities();
  platform([WORKER_APP_PLATFORM])
    .asyncApplication(setupIsolate(replyTo))
      .then((ref) => ref.bootstrap(RootComponent));
}

```

You should no longer import from the `angular2/web_worker/worker` and `angular2/web_worker/ui` paths. Instead you can now import directly from core, directives, etc..

The WebWorkerApplication class has been removed. If you want to use ServiceMessageBroker or ClientMessageBroker on the render thread, you must inject their factories via DI.
If you need to use the MessageBus on the render thread you must also obtain it through DI.

closes #3277
closes #5473

mary-poppins commented:
Merging PR #5519 on behalf of @jelbourn to branch [presubmit-jelbourn-pr-5519](https://github.com/angular/angular/tree/presubmit-jelbourn-pr-5519).

zhaxi commented:
@vsavkin @jteplitz602 webworkers can only be supportted in android 4.4+, If I want to support android 4.4- and optimize performance in android 4.4+ with webworkers, does it mean I have to write two sets of codes? could it be like "View Encapsulation" you guys invented, just one set of Code makes everyone happy!!
